paddle/phi/kernels/funcs/concat_and_split_functor.cu

```diff
@@ -19,111 +19,155 @@ limitations under the License. */
+    col_data = static_cast<IndexT*>(tmp_dev_ins_col_data->ptr());
+    data_warpper = PointerWarpper<T, 0>(
+        ctx, ins, static_cast<int64_t>(in_num), inputs_data);
+  }
+
+  __device__ inline const T* operator[](int i) const { return data_warpper[i]; }
+
+ private:
+  PointerWarpper<T, 0> data_warpper;
+};
+
+template <typename T, typename IndexT, typename WarpperT>
+__global__ void ConcatKernel_(WarpperT ins_datas,
                               int col_size,
-                              const int64_t output_rows,
-                              const int64_t output_cols,
+                              const IndexT output_rows,
+                              const IndexT output_cols,
                               T* output) {
-  int64_t curr_segment = 0;
-  int64_t curr_offset = input_cols[0];
-  CUDA_KERNEL_LOOP_TYPE(tid_x, output_cols, int64_t) {
-    int64_t curr_col_offset = input_cols[curr_segment + 1];
+  IndexT curr_segment = 0;
+  IndexT curr_offset = ins_datas.col_data[0];
+  CUDA_KERNEL_LOOP_TYPE(tid_x, output_cols, IndexT) {
+    IndexT curr_col_offset = ins_datas.col_data[curr_segment + 1];
     while (curr_col_offset <= tid_x) {
       curr_offset = curr_col_offset;
       ++curr_segment;
-      curr_col_offset = input_cols[curr_segment + 1];
+      curr_col_offset = ins_datas.col_data[curr_segment + 1];
     }
 
-    int64_t local_col = tid_x - curr_offset;
-    int64_t segment_width = curr_col_offset - curr_offset;
+    IndexT local_col = tid_x - curr_offset;
+    IndexT segment_width = curr_col_offset - curr_offset;
 
-    const T* input_ptr = inputs[curr_segment];
-    int64_t tid_y = blockIdx.y * blockDim.y + threadIdx.y;
+    const T* input_ptr = ins_datas[curr_segment];
+    IndexT tid_y = blockIdx.y * blockDim.y + threadIdx.y;
     for (; tid_y < output_rows; tid_y += blockDim.y * gridDim.y)
       output[tid_y * output_cols + tid_x] =
           input_ptr[tid_y * segment_width + local_col];
   }
 }
 
-template <typename T>
-__device__ void ConcatKernelDetail(const T** inputs_data,
-                                   const int64_t fixed_in_col,
-                                   const int64_t out_rows,
-                                   const int64_t out_cols,
-                                   T* output_data) {
-  CUDA_KERNEL_LOOP_TYPE(tid_x, out_cols, int64_t) {
-    int64_t split = tid_x * 1.0 / fixed_in_col;
-    int64_t in_offset = tid_x - split * fixed_in_col;
-    const T* input_ptr = inputs_data[split];
-    int64_t tid_y = blockIdx.y * blockDim.y + threadIdx.y;
+template <typename T, typename IndexT, typename WarpperT>
+__global__ void ConcatKernel(WarpperT ins_data,
+                             const IndexT fixed_in_col,
+                             const IndexT out_rows,
+                             const IndexT out_cols,
+                             T* output_data) {
+  CUDA_KERNEL_LOOP_TYPE(tid_x, out_cols, IndexT) {
+    IndexT split = tid_x * 1.0 / fixed_in_col;
```

Comment:
If possible, could ```1.0/fixed_in_col``` also be passed by kernel args?

Reply:
Good one, will be fixed in next commit.